diatheke/internal/audio/audio.go

```
@@ -209,3 +214,138 @@ func (p *Player) PushAudio(audio []byte) error {
+		origLen := sr.buffer.Len()
+		sr.buffer = *bytes.NewBuffer(sr.buffer.Bytes()[sr.buffer.Len()-sr.maxBufferSize:])
+		sr.bufferStartOffset += origLen - sr.maxBufferSize
+	}
+	return sr.currentReader.Read(p)
+}
+
+// Stop forces the next Read() of the StoppableReader to return EOF, but
+// additional Read() operations will continue to read bytes as usual.
+func (sr *StoppableReader) Stop() {
+	sr.mu.Lock()
+	sr.pauseRead = true
+	sr.mu.Unlock()
+}
+
+// Rewind returns a new Reader that starts at the specified byte offset.
+// This may allow data that had been previously read to be re-read.
+// If the StoppableAudioReader has a maximum buffer size and data
+// was not fully retained for the specified offset an error will be
+// returned, but a reader with as much buffered context as possible
+// will still be returned (what to do in this situation is up to
+// the caller).
+// if resetTimeZero is true, then future "offset" values sent to Rewing()
+// will consider the start of the last rewound stream to be zero, otherwise
+// the "zero time" will remain the start of the original input stream.
+func (sr *StoppableReader) Rewind(offset int, resetTimeZero bool) error {
+	sr.mu.Lock()
+	defer sr.mu.Unlock()
+
+	var err error
+	if sr.rewindWithoutReset {
+		// This is only occasionally an error (depending on how it is used). The caller may choose
+		// to ignore the error, but it is discouraged.
+		err = fmt.Errorf("Rewind() should not be called twice without a Reset() between them because reading a rewound buffer may modify the buffered data")
+	} else {
+		sr.rewindWithoutReset = true
+	}
+
+	adjustedOffset := offset
+	if offset < sr.bufferStartOffset {
+		// Offset value is before the start of the buffer, data returned will be incomplete.
+		adjustedOffset = 0
+		err = fmt.Errorf("StoppableAudioReader::Rewind() error, the requested offset %d is smaller that the start of the buffer: %d",
```

should we be returning this error here (like in the else-if case) ?

---

we do eventually return the error set in line 310 ... but before returning it we Rewind as far as possible.  When the caller sees that there is an error they can decide to handle or ignore the error (I think usually we'd ignore it, it's not that serious, mostly an indication that we should increase the buffer size unless the error is very uncommon or it was an unusual circumstance).